pkg/operator/controller/ca-configmap/controller_test.go

```diff
@@ -0,0 +1,213 @@
+*/
+
+package ca_configmap
+
+import (
+	"context"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	test "github.com/openshift/external-dns-operator/pkg/operator/controller/utils/test"
+)
+
+const (
+	testSrcConfigMapName    = "test-trusted-ca"
+	testTargetConfigMapName = "external-dns-trusted-ca"
+)
+
+func TestReconcile(t *testing.T) {
+	managedTypesList := []client.ObjectList{
+		&corev1.ConfigMapList{},
+	}
+	eventWaitTimeout := time.Duration(1 * time.Second)
+
+	testCases := []struct {
+		name            string
+		existingObjects []runtime.Object
+		inputConfig     Config
+		inputRequest    ctrl.Request
+		expectedResult  reconcile.Result
+		expectedEvents  []test.Event
+		errExpected     bool
+	}{
+		{
+			name:            "Bootstrap",
+			existingObjects: []runtime.Object{testSrcConfigMap()},
+			inputConfig:     testConfig(),
+			inputRequest:    testRequest(),
+			expectedResult:  reconcile.Result{},
+			expectedEvents: []test.Event{
+				{
+					EventType: watch.Added,
+					ObjType:   "configmap",
+					NamespacedName: types.NamespacedName{
+						Namespace: test.OperandNamespace,
+						Name:      testTargetConfigMapName,
+					},
+				},
+			},
+		},
+		{
+			name:            "Target configmap drifted",
+			existingObjects: []runtime.Object{testSrcConfigMap(), testDriftedConfigMap()},
+			inputConfig:     testConfig(),
+			inputRequest:    testRequest(),
+			expectedResult:  reconcile.Result{},
+			expectedEvents: []test.Event{
+				{
+					EventType: watch.Modified,
+					ObjType:   "configmap",
+					NamespacedName: types.NamespacedName{
+						Namespace: test.OperandNamespace,
+						Name:      testTargetConfigMapName,
+					},
+				},
+			},
+		},
+		{
+			name:            "Target configmap didn't change",
+			existingObjects: []runtime.Object{testSrcConfigMap(), testTargetConfigMap()},
+			inputConfig:     testConfig(),
+			inputRequest:    testRequest(),
+			expectedResult:  reconcile.Result{},
+		},
+		{
+			name:            "Deleted source configmap",
+			existingObjects: []runtime.Object{},
+			inputConfig:     testConfig(),
+			inputRequest:    testRequest(),
+			expectedResult:  reconcile.Result{},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			cl := fake.NewClientBuilder().WithScheme(test.Scheme).WithRuntimeObjects(tc.existingObjects...).Build()
+
+			r := &reconciler{
+				client: cl,
+				scheme: test.Scheme,
+				config: tc.inputConfig,
+				log:    zap.New(zap.UseDevMode(true)),
+			}
+
+			c := test.NewEventCollector(t, cl, managedTypesList)
```

Comment:
managedTypesList  so we everytime we pass managedTypesList [&ConfigMapList{ListMeta:{   <nil>},Items:[]ConfigMap{},}]  But I am not getting then how iteration happens on this list.

Reply:
`managedTypesList` is a list of types to be watched, we need only types without any items as they will be passed to `Watch()` function later in `EventCollector.Start()`: [link](https://github.com/openshift/external-dns-operator/blob/96aebfb5f7ac49e1eb849b48d82112e9c9e51885/pkg/operator/controller/utils/test/events.go#L114).

Reply:
Clarified this in Slack. The confusion was in the point that the watch type is a list object which is not used a container for items.